### ISSUE TITLE
validate-if-empty invalidates empty, not required fields

### ISSUE THREAD

**Issue by eradin** [guillaumepotier/Parsley.js#1151]
What kind of issue is this? (put 'x' between the square brackets)

 - [ ] Question. This issue tracker is not the place for questions. If you want to ask how to do
       something, or to understand why something isn't working the way you expect it to, use
       http://stackoverflow.com/questions/ask .
       Provide working code, starting from http://codepen.io/marcandre/pen/jqbzyN?editors=101.
       We monitor the tag `parsley.js`.

 - [ x] Bug report. If you’ve found a bug, you must provide a minimal example in a CodePen,
       starting from http://codepen.io/marcandre/pen/jqbzyN?editors=101 .

 - [ ] Feature Request. Make sure there's no good way to do what you want first;
       consider asking on http://stackoverflow.com/questions/ask first.

I think this is a bug as to opposed to an intended feature.  If validate-if-empty is set, field is not required, and the field is blank, the field is considered invalid.  A blank field should only be invalid if it's required in some way.  Here is a simple example using an email type field.  I didn't try all field types.  Perhaps this is only a problem with email.

http://codepen.io/eradin/pen/xRJXvL

**Comment by marcandre** [comment]
An empty string is not a valid email address, so it seems legitimate that it does not validate. If you change the input type to 'text', it validates properly.

**Comment by eradin** [comment]
That is preposterous.  A blank email is valid unless it's required.  The fact that a text field works as expected proves that point since an email type is just a variation of a text field.  Even chrome and firefox will not invalidate a blank non-required email.

Perhaps I don't understand the validate-if-empty attribute.  As I understand it, it simply runs the validator even though there is no input.  There are occasional reasons to do this but it shouldn't change the validation rules, it simply forces the validation method.  Non required email fields are valid (as well as any other text field).  If they weren't, they'd be required.

I hope you will revisit this.

**Comment by marcandre** [comment]
TBH, I know of no good use case of `validate-if-empty`. I have an even greater difficulty imagining a valid use case of `validate-if-empty` with `type=email`. Can you provide one?

Currently, one can check if a specific email is valid with:

    var isGoodEmail = function(val) {
      return $('<input type="email">').parsley().isValid({value: val, force: true});
    };

What you are suggesting, IIUC, is that it returns `isGoodEmail('') == true`. I don't see how this would be useful.

**Comment by henrythewasp** [comment]
My 2 cents:  I have a set of custom Parsley validators that do things 
like "conditionalrequired" - ie. this field is required if one or more 
other conditions are satisfied.  So you could use that on a form with a 
checkbox and an input field where the input field only needed to be 
validated if the checkbox was ticked.  I use validate-if-empty with the 
conditionalrequired validator to force Parsley to check it even when 
it's empty.

**Comment by marcandre** [comment]
I wish the interdependencies had been fixed :blush: 

So the proposal would be that all builtin validators accept `''` as a valid value?

**Comment by eradin** [comment]
@henrythewasp That would be my scenario as well.

@marcandre I believe that's correct.  Remember, anyone can alter that rule by simply adding required.

**Comment by henrythewasp** [comment]
@eradin I've also got a custom validator called conditionaltype, which runs a particular type validator if certain conditions are met - ie. this must be a valid email address if that checkbox is ticked.  That way we can do proper validation conditionally without need 'required' and so I don't need the built-in validators to accept '' as a valid value - in fact such a change would break our code.

**Comment by henrythewasp** [comment]
@marcandre Because I would want the email address to be a non-empty valid email address.  If user ticks checkbox and doesn't enter an email address then that needs to be caught by parsley.  So it's like a conditional required but with added type checking.   

If I understand the proposal, it would mean that an empty value is accepted as a valid email address (assuming the element doesn't have 'required' set).

I agree, we could use the disabled attr, but there are also other instances where we have multiple conditions and it's just been easier to apply those conditions with our custom validators.

**Comment by henrythewasp** [comment]
@marcandre Also - and this is probably down to my lack of understanding - it seems that fields that are excluded from Parsley cos they're disabled will not be validated if they are un-disabled unless I reload parsley.  I'm using 'excluded' field in a 'defaults' struct passed to parsley().  

Is there another way I can dynamically filter stuff during validation?

**Comment by marcandre** [comment]
> I would want the email address to be a non-empty valid email address. If user ticks checkbox and doesn't enter an email address then that needs to be caught by parsley

I'm sorry, I don't understand. Isn't this case already caught by your custom validator? Also, if the user doesn't tick the checkbox, the input is still `validate-if-empty`, and you wouldn't want an error to be generated, like the original requester?

> fields that are excluded from Parsley cos they're disabled will not be validated if they are un-disabled unless I reload parsley.

Not the case. Parsley constantly rescans for inputs by calling `_refreshFields`.

**Comment by robwilde** [comment]
I am getting the same issue with a number field as well as the email field.
So far what I found is:

`$this_field.parsley().validate() === true`
`$this_field.parsley().isValid() === true`
`$this_field.parsley().isValid({force: true}) === false`

the isValid() method calls whenValid() witch returns 'rejected' when the email and number fields are empty and not required.
parsley.js:1555

please if you need any more details.

**Comment by marcandre** [comment]
IIUC, my answer would be that an empty string is also not a valid number, so that the above is correct. If that's not what you meant, as usual *please provide actual working example*.

**Comment by robwilde** [comment]
Yes that's correct, form with a number field that is not required. It also appears to be the same issue with email field.
I take it the work around is to no use isValid({Force: true})

**Comment by marcandre** [comment]
> I take it the work around is to no use isValid({Force: true})

I agree there are very few instances where one wants to use in `force: true`, the same as `validate-if-empty` for anything else than some custom validators.

I understand that a combination of the builtin validators with the custom ones requiring `validate-if-empty` would be impossible though.

I'll change the builtin validators to accept empty strings then.

**Comment by marcandre** [comment]
Have a look at af2f99e, let me know if it looks good or not.

**Comment by benjaminpick** [comment]
Yes this fixes this issue for me ...

**Comment by marcandre** [comment]
This change has caused a few issues. I'm very tempted to simply revert it.

**Comment by benjaminpick** [comment]
If you revert it, do you have a different suggestion to solve this issue?

**Comment by paolohu** [comment]
This was not a good change - pretty much broke all our form validation had to roll back to the previous version we were using.
